Breezy Velvet Yak

Medium

# Missing withdrawalCallback call in `emergencyWithdraw()`

## Summary

Both `processWithdrawals()` and `emergencyWithdraw()` should call the withdraw callback.

## Vulnerability Detail

Using the `processWithdrawals()` function triggers the `withdrawalCallback()`. However, exiting using `emergencyWithdraw()` does not call the callback. This might result in situations where the callback is not synchronized with the vault.

## Impact

Although no withdrawal callback has been implemented, we can still measure its impact. Given the logic of synchronizing the callback with the vault, the impact could be significant.

## Code Snippet
[Vault.sol#L371-L416](https://github.com/sherlock-audit/2024-06-mellow/blob/main/mellow-lrt/src/Vault.sol#L371-L416)
```solidity
function emergencyWithdraw(
  uint256[] memory minAmounts,
  uint256 deadline
)
  external
  nonReentrant
  checkDeadline(deadline)
  returns (uint256[] memory actualAmounts)
{
  ...
  //@audit doesn't call the callback
  emit EmergencyWithdrawal(sender, request, actualAmounts);
}

function processWithdrawals(
  address[] memory users
) external nonReentrant returns (bool[] memory statuses) {
 ...
  IWithdrawalCallback(callback).withdrawalCallback();
  emit WithdrawCallback(callback);
}
```

## Tool used

Manual Review

## Recommendation

Itâ€™s recommended to include the call to the withdrawal callback